foundationdb/src/directory/mod.rs

```diff
@@ -0,0 +1,409 @@
+    /// (creating parent directories if necessary).
+    pub async fn create_or_open(
+        &self,
+        txn: &Transaction,
+        paths: Vec<String>,
+    ) -> Result<Subspace, DirectoryError> {
+        self.create_or_open_internal(txn, paths, vec![], true, true)
+            .await
+    }
+
+    /// Create creates a directory specified by path (relative to this
+    /// Directory), and returns the directory and its contents as a
+    /// Subspace (or ErrDirAlreadyExists if the directory already exists).
+    pub async fn create(&self, txn: &Transaction, paths: Vec<String>) -> Option<DirectoryError> {
+        self.create_or_open_internal(txn, paths, vec![], true, false)
+            .await
+            .err()
+    }
+
+    /// Open opens the directory specified by path (relative to this Directory),
+    /// and returns the directory and its contents as a Subspace (or Err(DirNotExists)
+    /// error if the directory does not exist, or ErrParentDirDoesNotExist if one of the parent
+    /// directories in the path does not exist).
+    pub async fn open(
+        &self,
+        txn: &Transaction,
+        paths: Vec<String>,
+    ) -> Result<Subspace, DirectoryError> {
+        self.create_or_open_internal(txn, paths, vec![], false, true)
+            .await
+    }
+
+    /// Exists returns true if the directory at path (relative to the default root directory) exists, and false otherwise.
+    pub async fn exists(
+        &self,
+        trx: &Transaction,
+        paths: Vec<String>,
+    ) -> Result<bool, DirectoryError> {
+        let nodes = self.find_nodes(trx, paths.to_owned()).await?;
+
+        match nodes.last() {
+            None => Err(DirectoryError::DirNotExists),
+            Some(_) => Ok(true),
+        }
+    }
+
+    /// Move moves the directory at oldPath to newPath (both relative to this
+    /// Directory), and returns the directory (at its new location) and its
+    /// contents as a Subspace. Move will return an error if a directory
+    /// does not exist at oldPath, a directory already exists at newPath, or the
+    /// parent directory of newPath does not exist.
+    pub async fn move_to(
+        &self,
+        trx: &Transaction,
+        old_path: Vec<String>,
+        new_path: Vec<String>,
+    ) -> Result<bool, DirectoryError> {
+        self.check_version(trx, false).await?;
+
+        let mut old_nodes = self.find_nodes(&trx, old_path.to_owned()).await?;
+        let last_node_from_old_path = match old_nodes.last_mut() {
+            None => return Err(DirectoryError::Message(String::from("empty path"))),
+            Some(node) => node,
+        };
+        let content_subspace = last_node_from_old_path
+            .content_subspace
+            .as_ref()
+            .ok_or(DirectoryError::DirNotExists)?;
+
+        let mut new_nodes = self.find_nodes(&trx, new_path.to_owned()).await?;
+        // assert that parent of the new node exists
+        if new_nodes.get(new_nodes.len() - 2).is_none() {
```

Comment:
panic if `len() < 2`

Reply:
You are right, I backported the code from the Go binding, and I'm wondering why we cannot move on the root-node. I'll definitely add more restrictions :smile